_source/_posts/2022-09-26-java-jetty-oauth.md

```diff
@@ -0,0 +1,1017 @@
+            return "Param 'hike' cannot be null.";
+        }
+        else if (this.hikes.contains(hike)) {
+            response.setStatus(400);
+            return "The hike '"+hike+"' already exists.";
+        }
+        else {
+            this.hikes.add(hike);
+            return String.join("\n", this.hikes);
+        }
+    }
+
+    @DeleteMapping("")
+    @ResponseBody
+    @PreAuthorize("isAuthenticated")  // <- ***ADDED***
+    public String indexDelete(@RequestParam String hike, HttpServletResponse response) {
+        if (hike == null) {
+            response.setStatus(400);
+            return "Param 'hike' cannot be null.";
+        }
+        
+        
+        else {
+            this.hikes.remove(hike);
+            return String.join("\n", this.hikes);
+        }
+    }
+
+}
+```
+
+Try to POST a new hike.
+
+```bash
+http -f POST :8080 hike="Pear Lake"
+```
+
+```bash
+HTTP/1.1 403 Forbidden
+...
+
+{
+    "error": "Forbidden",
+    "path": "/",
+    "status": 403,
+    "timestamp": "2022-09-06T14:09:02.533+00:00"
+}
+```
+
+## Generate a JWT using the OIDC debugger
+
+To access the protected endpoints, you need to generate an access token JWT. {% include setup/oidcdebugger.md %}
+
+{% img blog/java-jetty/oidc-debugger.png alt:"OIDC Debugger Configuration" width:"650" %}{: .center-image }
+
+Scroll down and click **Send Request**.
+
+This will respond with an authorization code, which you need to send back to the authorization URI to retrieve the code.
+
+```bash
+http -f POST https://dev-447850.okta.com/oauth2/default/v1/token \
+  grant_type=authorization_code \
+  code=i3sP86Ru7spu8gOQUBZUtNC7yBnCQ8PggxJAIBMySpE \
+  client_id=0oa918jqqmcbBZxMG4x7 \
+  client_secret=VJ8K9k0dZ-pDl9d1BaY2ibaexpFv3vc99OtuHndq \
+  redirect_uri=https://oidcdebugger.com/debug
+```
+
+```bash
+HTTP/1.1 200 OK
+...
+
+{
+  "access_token": "eyJraWQiOiJqY3dpbGpUcGVZSG1Jajl6ODR3LVVLQm...",
+  "expires_in": 3600,
+  "id_token": "eyJraWQiOiJqY3dpbGpUcGVZSG1Jajl6ODR3LVVLQmthYm...",
+  "scope": "openid",
+  "token_type": "Bearer"
+}
+```
+
+Copy the access token to your clipboard and store it in a shell variable in the shell window you use to make requests.
+
+```
+TOKEN=eyJraWQiOiJIb05xb01mNE9jREltWnBGRnBINjZGTkFOM0J... 
+```
+
+Now try and POST a new hike and then remove it.
+
+```bash
+http -f POST :8080 hike="Pear Lake" "Authorization: Bearer $TOKEN"
+```
+
+```
+HTTP/1.1 200 OK
+...
+
+Wonderland Trail
+South Maroon Peak
+Tour du Mont Blanc
+Teton Crest Trail
+Everest Base Camp via Cho La Pass
+Kesugi Ridge
+Pear Lake
+```
+
+```bash
+http DELETE :8080 hike=="South Maroon Peak" "Authorization: Bearer $TOKEN"
+```
+
+```
+HTTP/1.1 200 OK
+...
+
+Wonderland Trail
+Tour du Mont Blanc
+Teton Crest Trail
+Everest Base Camp via Cho La Pass
+Kesugi Ridge
+Pear Lake
+```
+
+## Secure the Spring Boot app with Auth0
+
+You can also use Auth0 to secure the Spring Boot application. Look at the project in the `spring-boot-jetty-maven-auth0` subdirectory.
```

Comment:
I like the idea of showing both Okta and Auth0 and having two separate examples on GitHub. 👏